enginetest/queries/insert_queries.go

```diff
@@ -1426,6 +1426,58 @@ var InsertScripts = []ScriptTest{
 			},
 		},
 	},
+	{
+		Name: "Insert on duplicate key references table in subquery",
+		SetUpScript: []string{
+			`create table a (i int primary key)`,
+			`insert into a values (1)`,
+			`create table b (j int primary key)`,
+			`insert into b values (1), (2), (3)`,
+		},
+		Assertions: []ScriptTestAssertion{
+			{
+				Query: `insert into a (select * from b) on duplicate key update a.i = b.j + 100`,
```

Comment:
Would be good to have a test with an aliased table name for `b`. I just tested that on MySQL and a table alias did resolve correctly. I also tested if a column alias defined in the source would be visible in onDupKey for MySQL, but it doesn't seem to be from what I could see.

Reply:
Probably also worth having some of those error cases. Particularly for the case where GMS was updating the wrong table – would be great to include that statement and make sure it errors out.